File: common/src/main/java/generations/gg/generations/core/generationscore/world/shop/ShopPreset.java
package generations.gg.generations.core.generationscore.world.shop;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.network.FriendlyByteBuf;

import java.util.List;

public record ShopPreset(List<ShopEntry> entries, int maxAppearingItems) {
    public ShopPreset(FriendlyByteBuf buf) {
        this(buf.readList(ShopEntry::decode), buf.readInt());
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeCollection(entries, (buf1, shopEntry) -> shopEntry.encode(buf1));
        buf.writeInt(maxAppearingItems);
    }

    public static Codec<ShopPreset> CODEC = RecordCodecBuilder.create(instance -> instance.group(
            Codec.list(ShopEntry.CODEC).fieldOf("entries").forGetter(ShopPreset::entries),
            Codec.INT.fieldOf("maxAppearingItems").forGetter(ShopPreset::maxAppearingItems)
    ).apply(instance, ShopPreset::new));
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/shop/ShopPreset.kt
package generations.gg.generations.core.generationscore.common.world.shop

import com.cobblemon.mod.common.util.writeCollection
import com.cobblemon.mod.common.util.writeItemStack
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import generations.gg.generations.core.generationscore.common.world.shop.ShopPreset
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.resources.ResourceLocation

data class ShopPreset(val id: ResourceLocation, val entries: List<ShopEntry>, val maxAppearingItems: Int) {
    companion object {
        fun decode(buf: RegistryFriendlyByteBuf): ShopPreset {
            val id = buf.readResourceLocation()
            val entries = ShopEntry.LIST_STREAM_CODEC.decode(buf)
            val maxAppearingItems = buf.readVarInt()
            return ShopPreset(id, entries, maxAppearingItems)
        }
    }

    fun encode(buf: RegistryFriendlyByteBuf) {
        buf.writeResourceLocation(id)
        ShopEntry.LIST_STREAM_CODEC.encode(buf, entries)
        buf.writeVarInt(maxAppearingItems)
    }
}